Implement TeX's fraction alignment for Mathtext fractions

tfpf opened the pull request:
## PR Summary
Follow-up to https://github.com/matplotlib/matplotlib/pull/20627, which somehow got closed automatically when I was messing around with merge conflicts in my fork. I have implemented the same algorithm TeX uses to align fractions.

At the moment, I have not updated the images, for reasons I will mention in a comment.

Fixes https://github.com/matplotlib/matplotlib/issues/18086. Fixes https://github.com/matplotlib/matplotlib/issues/18389. Fixes https://github.com/matplotlib/matplotlib/issues/22172.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

tfpf commented:
A few notes.

### Algorithm
To best of my knowledge, I tried to use the algorithm described in _TeX: The Program_ by Don Knuth. It describes numerators and denominators as being shifted independently of each other, whereas Matplotlib packs them (along with a rule vertical spaces) into a box, and moves the box down. As a result, some acrobatics were necessary, so there is scope of optimisation, but I focused on getting it to work first.

![with_rule](https://user-images.githubusercontent.com/19171016/163554737-26ef0c72-d3b5-426d-81ab-d7f8333d09c7.png)
Fractions with a rule. Top is text style. Bottom is display style.

![without_rule](https://user-images.githubusercontent.com/19171016/163555330-25abbece-5eac-467d-b85c-1ad84a393fbb.png)
Fractions without a rule. Top is text style. I couldn't figure out how to use `\genfrac` to create fractions in display style. If someone can tell me how to do that, I will test display style fractions without a rule and update this image.

And yes, the spacing does take into account the thickness of the rule, as TeX does.

### Font Constants
In order to use the same algorithm, I defined five new constants in the class `FontConstantsBase`, and used them to mean the same thing TeX does. (On an unrelated note, the comments describing these constants refer to them as 'percentages', but I think they should be called 'fractions', since they are direct multipliers. Should the comments be changed?)

I chose values which best rendered fractions while using DejaVu Sans. Different values may be needed for the other four typefaces.

### Possible Bug in Fraction Rendering
            num_clr = max(
                num_shift_up - cnum.depth - axis_height - rule / 2, min_clr)
            den_clr = max(
                axis_height - rule / 2 + den_shift_down - cden.height, min_clr)
            vlist = Vlist([cnum,                     # numerator
                           Vbox(0, num_clr - rule),  # space
                           Hrule(state, rule),       # rule
                           Vbox(0, den_clr + rule),  # space
                           cden                      # denominator
                           ])
This is the block of code which defines the clearance between the rule and the numerator/denominator. `num_clr` is the vertical distance between the numerator and the rule. `den_clr` is that between the denominator and the rule. If a `Vlist` is created with `num_clr` and `den_clr`, the actual rendered distances turn out to be `num_clr + rule`and `den_clr - rule`. This looks like a bug to me, so I undid its effects by specifying `num_clr - rule` and `den_clr + rule`. The actual rendered distances then become `num_clr - rule + rule = num_clr` and `den_clr + rule - rule = den_clr`, which matches what TeX would do.

If the bug gets fixed in the future, all we'd need to do is remove ` - rule` and ` + rule` from the above code block.

### Updating Images
The choice of font constants may work for DejaVu Sans, but probably won't work for the others. I think images should be added only after we have a consensus on what those values should be.

timhoffm commented on lib/matplotlib/_mathtext.py:
```diff
@@ -767,6 +767,23 @@ class FontConstantsBase:
+
+    # Percentage of x-height the numerator is shifted up in text style if there
+    # is a fraction line.
+    num2 = 1.5
```
Are these names used in TeX? If not, can we have slightly more descriptive names? e.g. `num_disp`, `num_text_frac`, `num_text_nofrac`

tfpf replied:
Yes, these are the names used in TeX. We already have `subdrop`, `sup1`, `sub1` and `sub2` (which are also the names of TeX font constants) in `FontConstantsBase` so I ran with it and used TeX's names (`num1`, `num2`, `num3`, `denom1` and `denom2`) for the new constants.

timhoffm reviewed (commented):
If you have it, including a reference to the TeX algorithm would be great.

tfpf commented:
> If you have it, including a reference to the TeX algorithm would be great.

Sure. I'll add a comment with the appropriate references.

tfpf commented:
I have added font constants for STIX, STIX sans, DejaVu Serif and CM. The values chosen are the nearest multiples of 0.1 for which the numerator and denominator seen in [the images in this comment](https://github.com/matplotlib/matplotlib/pull/22852#issuecomment-1100007576) do not have to be shifted beyond their default shift amounts.

anntzer commented:
Sorry to hijack this PR (which looks nice, although I haven't checked the exact algorithms yet), but perhaps we can revisit #19201 (smaller svg baselines by not embedding fonts) as well? (at least the part switching the svg baselines to not embedding the fonts, for a significant gain in size, even if we keep the png and pdf baselines still around for now.)  #19201 was deferred on the expectation of switching to a completely separate baseline images repo, but I'm not sure this is going to happen soon...
(I can extract the relevant parts of #19201 into a new PR, of course.)

tfpf commented:
That is a good idea (as I understand, fewer changes to the images will result in a smaller size increase of the repo), but I am really not sure how to do it. Perhaps someone else could push commits here?

tfpf commented:
Noticed something while fixing the merge conflict. Mathtext adds a space to the right of each fraction.

        result = [Hlist([vlist, Hbox(thickness * 2.)])]

Is there a reason for this? Fractions look better without that space:
![without_space](https://user-images.githubusercontent.com/19171016/164454135-5d78bcfd-8fd9-496b-8dfd-b06e3826eb98.png)

than with it.
![with_space](https://user-images.githubusercontent.com/19171016/164454172-dcf50d25-981c-4293-8456-a09157b1db82.png)

anntzer commented:
Agreed it looks better without the shift; a quick git blame suggests this comes from the original implementation (c10b4bbe) with no reason given :-(

anntzer commented:
I put up a revival of #19201 at #22881, let's discuss that there.

tfpf commented:
> from the original implementation with no reason given

That is unfortunate.😐 At a glance, it appears that this space was added so that the lines of two consecutive fractions (e.g. `\dfrac{1}{2}\dfrac{1}{2}`) do not look like a single line. But that raises the question why the space was not added on both sides like this:

        result = [Hlist([Hbox(thickness * 1.), vlist, Hbox(thickness * 1.)])]

instead of just on the right.

Although it's just a single change, I think this deserves a separate PR?

anntzer commented:
Seems like a reasonable guess as to why this came in.  Do you have any idea of what TeX does to avoid the same problem?

tfpf commented:
From a quick look at the book, node 748 seems to suggest that the `Hlist` itself does not contain any extra space. Perhaps a space is added only if there is something before or after the fraction? I have no clue.

tfpf commented:
> Do you have any idea of what TeX does to avoid the same problem?

I think I've got it. Per node 764 of the TeX book, all possible inter-noad space codes are first stored in a 64-element string, which is treated as if it were an 8-by-8 matrix. The space between two noad types `a` and `b` is computed using the element in the `a`th row and `b`th column of the matrix, along with what is called a `magic_constant`. (I think that's the general idea.)

Here, `a` and `b` are numbers representing the noad type. A noad is an element of an mlist (which is a linked list representing formulae). Hence, the space between two fractions is `8 * fraction_noad + fraction_noad + magic_offset`th element of the matrix.

Is this beyond the scope of Mathtext? (Will there be any side-effects of making [this change](https://github.com/matplotlib/matplotlib/pull/22852#issuecomment-1108191351).)

jklymak commented:
Moving to draft.  Feel free to move back when the relevant tests pass!

anntzer commented:
@tfpf I think best would be to perhaps not encode the entire 64 internoad spaces for now (unless you want to go for it) and just specifically encode an inter-fraction noad?  Then perhaps squash this together with #21850 and accept the mass changes in baseline images (although I would **much** prefer if #22881 could go in first, too, and switch all the changed images to svg-unembedded-glyphs-only).

tfpf commented:
> just specifically encode an inter-fraction noad

As far as I can tell, Mathtext's spacing algorithm is not the same as that of TeX. Mathtext adds spaces to `Hlist` objects. I think it might be simpler to just put equal spaces on both sides of the fraction.

> switch all the changed images to svg-unembedded-glyphs-only

Agreed, that is the most meaningful thing to do. I will hold off on changing any images until that pull request is merged.

tfpf commented:
>  Then perhaps squash this together

@anntzer Does this mean: add the subscript and superscript changes here? This would result in one less change for some images, because there are some tests which have fractions as well as superscripts and subscripts.

However, I had to cut some corners for the latter. I'll paste the details here.

> ### Shortcuts I took for the algorithm
> - The default values of `shift_up` and `shift_down` depend on two font constants, `supdrop` and `subdrop`, respectively. However, Mathtext uses `subdrop` for both (as the comment in the `FontConstantsBase` class mentions), so I did the same.
> - The actual value of `shift_up` depends on the current style (display, text, script or script-script), and may be calculated using `sup1`, `sup2` or `sup3`. However, superscripts and subscripts in Mathtext do not know about the current style (Mathtext only has the `sup1` constant); they are simply one size smaller than the nucleus. I've also used only `sup1`.

anntzer commented:
> Does this mean: add the subscript and superscript changes here? This would result in one less change for some images, because there are some tests which have fractions as well as superscripts and subscripts.

Yes.

> Also, if we are to use the TeX algorithm for superscripts and subscripts, the "poor man's x-height" calculation (#21850 (comment)) may have to be forced for all fonts. (If there is a better way, I am all for it, though.)

IIRC(?) there are other places where we noted that font-reported-x-heights can be useless and it may be better to use poor man's x-height everywhere instead, so I wouldn't be too annoyed by that change.

-----

The shortcuts you mention seem fine.

tfpf commented:
Of the 537 Mathtext image comparison tests failed, there were some that failed for Computer Modern, but not for the other fonts. I moved all the failing tests to `svgastext_math_tests`, but since each test checks all five fonts, a total of 720 new images have to added.:sweat_smile:

There are a few errors I don't quite understand.
- `r'$\sqrt{1+\sqrt{1+\sqrt{1+\sqrt{1+\sqrt{1+\sqrt{1+\sqrt{1+x}}}}}}}$'` (There are neither fractions nor superscripts or superscripts. And yet, this image failed the image comparison test.)
- `r"$f'\quad f'''(x)\quad ''/\mathrm{yr}$"` (Looks like apostrophes affected by the superscript positioning.)
- `test_operator_space` (In this `check_figures_equal` test, the 's' of 'cos' is displaced slightly compared to the reference image, even though 'co' is placed identically. This is strange, because operator kerning should have remained unchanged.)

Also, I can't quite figure out how to identify the old images to be deleted.

tfpf commented:
> Of the 537 Mathtext image comparison tests failed, there were some that failed for Computer Modern, but not for the other fonts. I moved all the failing tests to svgastext_math_tests, but since each test checks all five fonts, a total of 720 new images have to added.:sweat_smile:

This would mean that some images can simply be renamed instead of being deleted and added back … if one could sift through all of them. All-in-all, this is a probably too big a change, so I'll be happy to close it if it doesn't work out.

anntzer commented:
Sorry, I slightly messed up the "svg lightweight tests" change.  Can you additionally include the two lines at https://github.com/matplotlib/matplotlib/pull/23270/files#diff-17f9692b77108a62b59c71825d8461488729ef280f56785d3ad55ce3a86044d8R210 to your PR?  This way only svg will be tested, and only for dejavusans and cm; you can then remove the other baselines.

I don't think(?) there's a really smart way to figure out what images can be deleted other than by manually checking...  (Probably it would be nice to have some tooling for that.)

tfpf commented:
Sure, I will update it.

tfpf commented:
Will rebase since https://github.com/matplotlib/matplotlib/pull/23270 has been merged.

tfpf commented:
`test_operator_space` has me stumped!

---
### `test_operator_space[png].png`
![test_operator_space png](https://user-images.githubusercontent.com/19171016/175097935-c937365c-6267-4f8b-bf79-815ad79178ed.png)

---
### `test_operator_space[png]-expected.png`
![test_operator_space png -expected](https://user-images.githubusercontent.com/19171016/175097941-16eaa3a1-fd74-4027-93ca-f9efe88c5f13.png)

---
### `test_operator_space[png]-failed-diff.png`
![test_operator_space png -failed-diff](https://user-images.githubusercontent.com/19171016/175097944-e2418fe2-30ab-4b74-b959-3047d441ed49.png)

---
The only difference between the figures is that the space between the 'o' and 's' of 'cos' is more in one case than the other. The code changes shouldn't have affected inter-character kerning.

`\cos` is rendered as
```
Hlist<w=24.61 h=8.00 d=0.00 s=0.00>[`c`, k0.90, `o`, k0.96, `s`, k1.07, k2.31]
```
while `\mathrm{\,cos}` rendered as
```
[Hlist<w=24.61 h=8.00 d=0.00 s=0.00>[k2.31, `c`, k0.90, `o`, k0.96, `s`, k1.07]]
```
which clearly shows that the space between 'o' and 's' is identical. How do the renders then end up looking different.:confused:

anntzer commented:
My (very not guaranteed...) understanding is the following:
Looking at the hlists passed to ship(), the old version was (after stripping out Hlist sizes)
```
(A) H[H[], H[H[H[`s`, k1.07, `i`, k0.86, `n`], H[k0.74, `2`, k1.20], k2.44], H[`c`, k0.90, `o`, k0.96, `s`, k1.07, k2.31]], H[]]  # r"$\sin^2 \cos$"
(B) H[H[], H[H[H[`s`, k1.07, `i`, k0.86, `n`], H[k0.74, `2`, k1.20]], H[k2.31, `c`, k0.90, `o`, k0.96, `s`, k1.07]], H[]]  # r"$\mathrm{sin}^2\mathrm{\,cos}$"
```
and the new one is
```
(C) H[H[], H[H[H[`s`, k1.07, `i`, k0.86, `n`], H[k0.42, `2`, k1.20], k2.44], H[`c`, k0.90, `o`, k0.96, `s`, k1.07, k2.31]], H[]]
(D) H[H[], H[H[H[`s`, k1.07, `i`, k0.86, `n`], H[k0.42, `2`, k1.20]], H[k2.31, `c`, k0.90, `o`, k0.96, `s`, k1.07]], H[]]
```
i.e. the kern before the "2" changed.  (Why?)
Also, note that the kerns before the "c" are actually different in the two supposedly identical examples (is that a bug?)
This means that e.g., taking all positions relative to the "n" of "sin", the letter "c" is positioned in case (A) at round(0.74+widthof(2)+1.20+2.44) and in case (B) at round(0.74+widthof(2)+1.20+2.31) [the rounding is due to hinting, as glyphs can't really be drawn between two pixels (well, technically this depends on the hinting level)], and we're lucky that the rounding gets rid of the difference between 2.44 and 2.31.  OTOH, case (C) is round(0.42+widthof(2)+1.20+2.44) and case (D) is 
round(0.42+widthof(2)+1.20+2.31), and we may well not be as lucky with the rounding avoiding the change for us.

tfpf commented:
Thanks a lot for the patient and detailed analysis! I think I miscalculated the font constant updates for CM. Further, I hadn't changed `delta` to account for the changed x-height (as indicated by the kerning before '2'). If I fix those two things, maybe it'll work.

> Also, note that the kerns before the "c" are actually different in the two supposedly identical examples (is that a bug?)

Definitely looks like one. Might be worth a closer look.

anntzer commented on lib/matplotlib/_mathtext.py:
```diff
@@ -759,16 +754,38 @@ class FontConstantsBase:
-    sup1 = 0.45
-    sub1 = 0.2
-    sub2 = 0.3
+    script_space = 0.1476237
```
Can you comment on where these numbers come from?

tfpf replied:
Increased the constants by the same factor the x-height was decreased as a result of calculating the correct x-height, hoping that doing so would minimise the number of changes. (The details are in 70959c36dc7b4f07f15fb5bb82b1c3d054eb4b5d commit message, but yes, I'll be sure to mention the same in a comment as well, since I calculated the multiplying factor incorrectly and will force-push it.)

tfpf commented:
> Also, note that the kerns before the "c" are actually different in the two supposedly identical examples (is that a bug?)

The space before 'c' is `\,`, which is 1/6 em. Looking at the definition of `TruetypeFonts._get_info`, it looks like `k2.44` and `k2.31` are both 1/6 em, but the former uses an italicised 'm', whereas the latter uses a regular 'm'.

> the kern before the "2" changed. (Why?)

This can be fixed by increasing the CM constants by the correct multiplying factor, but the problem is that the multiplying factor depends on the ratio of the old x-height and the actual x-height (as returned by `get_metrics`). The actual x-height, in turn, depends on the hinting type. The function `TruetypeFonts._get_info` loads the metrics using the hinting type. I am not clear on what that exactly is, but if the hinting type is 2, the x-height of CM at font size 12 with 100 DPI is 7.375, whereas if the hinting type is 32, the x-height is 8 at the same font size and DPI.

I will go ahead and use the latter to calculate the multiplying factor, since that fixes `test_operator_space` and uses the same kerning as earlier.

tfpf commented on lib/matplotlib/tests/test_mathtext.py:
```diff
@@ -18,57 +18,123 @@
-    '$\\alpha \\beta \\gamma \\delta \\epsilon \\zeta \\eta \\theta \\iota \\lambda \\mu \\nu \\xi \\pi \\kappa \\rho \\sigma \\tau \\upsilon \\phi \\chi \\psi$',
 
     # The examples prefixed by 'mmltt' are from the MathML torture test here:
-    # https://developer.mozilla.org/en-US/docs/Mozilla/MathML_Project/MathML_Torture_Test
```
This link is dead. I have updated it.

tfpf commented on lib/matplotlib/tests/test_mathtext.py:
```diff
@@ -18,57 +18,123 @@
     r'$x+{y}^{\frac{2}{k+1}}$',
     r'$\frac{a}{b/2}$',
     r'${a}_{0}+\frac{1}{{a}_{1}+\frac{1}{{a}_{2}+\frac{1}{{a}_{3}+\frac{1}{{a}_{4}}}}}$',
-    r'${a}_{0}+\frac{1}{{a}_{1}+\frac{1}{{a}_{2}+\frac{1}{{a}_{3}+\frac{1}{{a}_{4}}}}}$',
```
Are we sure that the tests in `svgastext_math_tests` are being run correctly? I changed this test (which was a duplicate of the previous test) to use `dfrac` instead of `frac` (which is what the MathML test uses), but forgot to update the image file, since these tests were skipped on my system despite having Inkscape installed and on the path.

Shouldn't there have been test failures? Why are the tests passing?

anntzer replied:
Let's figure out #23350 first.

tfpf commented:
PR cleanliness failed.

If these changes are acceptable, I will squash the commits.

anntzer commented:
Previously, the two tests
```python
    r'$xyz^kx_kx^py^{p-2} d_i^jb_jc_kd x^j_i E^0 E^0_u$',  # github issue #4873
    r'${xyz}^k{x}_{k}{x}^{p}{y}^{p-2} {d}_{i}^{j}{b}_{j}{c}_{k}{d} {x}^{j}_{i}{E}^{0}{E}^0_u$',
```
rendered the same.

With this PR, some of the exponents are shifted between the two cases.  I think this is not desirable?

tfpf commented:
Probably a result of the difference between `{something}` and `something`, as discussed in https://github.com/matplotlib/matplotlib/issues/23257. I'll try to figure it out.

tfpf commented on lib/matplotlib/_mathtext.py:
```diff
@@ -2222,6 +2256,16 @@ def subsuper(self, s, loc, toks):
             return [result]
 
+        # Set the minimum shifts for the superscript and subscript (node756).
+        constants = _get_font_constant_set(state)
+        if isinstance(nucleus, Char):
+            shift_up = 0
+            shift_down = 0
+        else:
+            drop_amount = constants.subdrop * xHeight * SHRINK_FACTOR
+            shift_up = nucleus.height - drop_amount
+            shift_down = nucleus.depth + drop_amount
```
The problem comes from here. `{xyz}^k` triggers the `else` branch, while `xyz^k` triggers the `if` branch. This is consistent with node 756 of the book. However, TeX produces the same results for both. This is likely the result of extremely careful choices of font constants by Knuth. I can patch this by deleting the `if` branch, which should ensure that `{xyz}^k` and `xyz^k` get rendered identically.

tfpf commented:
This time, I double-checked that `x^2_y` and `x_y^2` are rendered identically, as are `xyz^kx_kx^py^{p-2} d_i^jb_jc_kd x^j_i E^0 E^0_u` and `{xyz}^k{x}_{k}{x}^{p}{y}^{p-2} {d}_{i}^{j}{b}_{j}{c}_{k}{d} {x}^{j}_{i}{E}^{0}{E}^0_u`.

Another thing is that `{a}_{0}+\dfrac{1}{{a}_{1}+\dfrac{1}{{a}_{2}+\dfrac{1}{{a}_{3}+\dfrac{1}{{a}_{4}}}}}` gets cut off at the top and bottom (the size of the figure isn't enough accommodate everything).

QuLogic commented:
There are a _lot_ of images changed here, but because they're renamed, it's difficult to really see what effect this change has on any of them. And the ones that are left with a diff sometimes look worse...

Plus some of the added images just don't look right at all:
![image](https://user-images.githubusercontent.com/302469/183803953-25571404-6656-4d73-89a9-50efa1a149f0.png)
(presumably due to missing fonts locally?)

If we want to combine this change with switch to SVG text, then we can do that _after_ review, or just in an entirely separate commit/PR.

tfpf commented:
> (presumably due to missing fonts locally?)

This is probably the reason, since Inkscape is able to correctly convert the SVG images to PNG. For instance, this is `mathtext0_cm_07_svg.png`.
![image](https://user-images.githubusercontent.com/19171016/183820296-af22bdea-a6b6-4990-9ffc-5c813093bdc1.png)

> If we want to combine this change with switch to SVG text, then we can do that after review, or just in an entirely separate commit/PR.

I think moving to SVG in a separate PR will be easier to manage.

anntzer commented:
The fact that the svg cm misrenders on a webbrowser is normal: this is because it is specifically tailored to use matplotlib's cm font files, which are the "classical" ones from AMS and use a nonstandard encoding: unless you specifically point the the right font files (which our _SVGConverter does), you won't get the right glyphs.  So the idea is really to just look at the dejavu svgs; if you want to confirm that the cm svgs are correct you should run the tests and visualize the files converted to png in result_images/.

tfpf commented:
> > Also, note that the kerns before the "c" are actually different in the two supposedly identical examples (is that a bug?)
> 
> The space before 'c' is \,, which is 1/6 em. Looking at the definition of TruetypeFonts._get_info, it looks like k2.44 and k2.31 are both 1/6 em, but the former uses an italicised 'm', whereas the latter uses a regular 'm'.
> 
> > the kern before the "2" changed. (Why?)
> 
> This can be fixed by increasing the CM constants by the correct multiplying factor, but the problem is that the multiplying factor depends on the ratio of the old x-height and the actual x-height (as returned by get_metrics). The actual x-height, in turn, depends on the hinting type. The function TruetypeFonts._get_info loads the metrics using the hinting type. I am not clear on what that exactly is, but if the hinting type is 2, the x-height of CM at font size 12 with 100 DPI is 7.375, whereas if the hinting type is 32, the x-height is 8 at the same font size and DPI.

I am still unable to figure this out, after all this time. Here's an example which demonstrates the problem. I made the following changes:
```diff
diff --git a/lib/matplotlib/_mathtext.py b/lib/matplotlib/_mathtext.py
index 3a934c21fd..93df6c4f05 100644
--- a/lib/matplotlib/_mathtext.py
+++ b/lib/matplotlib/_mathtext.py
@@ -334,6 +334,10 @@ class TruetypeFonts(Fonts):
         font = self._get_font(fontname)
         font.set_size(fontsize, dpi)
         pclt = font.get_sfnt_table('pclt')
+        print(f'{fontname=}, {fontsize=}', end=', ')
+        if pclt is not None:
+            print('Reported:', (pclt['xHeight'] / 64.0) * (fontsize / 12.0) * (dpi / 100.0), end=', ')
+        print('Actual:', self.get_metrics(fontname, mpl.rcParams['mathtext.default'], 'x', fontsize, dpi).iceberg)
         if pclt is None:
             # Some fonts don't store the xHeight, so we do a poor man's xHeight
             metrics = self.get_metrics(
```
and created the file `test.py`.
```
import matplotlib.pyplot as plt
plt.rcdefaults()
plt.rc('mathtext', fontset='cm')
plt.rc('font', size=12)
plt.figure().text(.5, .5, r'$x_1^2$')
plt.show()
```

Now comes the strange part.
```
$ python3 test.py 
fontname='it', fontsize=12.0, Reported: 14.171875, Actual: 8.0
$ pytest -s lib/matplotlib/tests/test_mathtext.py::test_mathtext_rendering[png-mathtext-cm-12]
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-7.1.2, pluggy-1.0.0
rootdir: /media/tfpf/OS1/Users/vpaij/Documents/projects/matplotlib, configfile: pytest.ini
plugins: xdist-2.5.0, forked-1.4.0
collected 1 item                                                               

lib/matplotlib/tests/test_mathtext.py fontname='it', fontsize=12.0, Reported: 14.171875, Actual: 7.375
fontname='it', fontsize=12.0, Reported: 14.171875, Actual: 7.375
.

============================== 1 passed in 0.77s ===============================
```

The x-height changes, presumably because the hinting is different in the tests. (Also, the ratio of the reported x-height height to the actual x-height of CM is not constant across font sizes.) I'll shelve this for a time I am better equipped to fix it.
